R/functions_degs.R

```diff
@@ -106,37 +157,509 @@ DegsWriteToFile = function(degs, annot_ensembl, file) {
+          }
+          contrast[["subset_group"]] = JoinSpecificationList(subset_group, first_level_separator=";", second_level_separator="+")
+        }
+      }
+    }
+    
+    # test
+    if (!"test" %in% names(contrast) || is.na(contrast[["test"]])) contrast[["test"]] = "wilcox"
+    if (!contrast[["test"]] %in% c("wilcox", "bimod", "roc", "t", "negbinom", "poisson", "LR", "MAST", "DESeq2")) error_messages = c(error_messages, paste0("The 'test' column value '",contrast[["test"]], "' in row ", i, " of the deg contrasts table must be one of: 'wilcox', 'bimod', 'roc', 't', 'negbinom', 'poisson', 'LR', 'MAST', 'DESeq2'."))
+    
+    # padj
+    if (!"padj" %in% names(contrast) || is.na(contrast[["padj"]])) contrast[["padj"]] = 0.05
+    
+    # log2FC
+    if (!"log2FC" %in% names(contrast) || is.na(contrast[["log2FC"]])) contrast[["log2FC"]] = 0
+    
+    # min_pct
+    if (!"min_pct" %in% names(contrast) || is.na(contrast[["min_pct"]])) contrast[["min_pct"]] = 0.1
+    
+    # assay
+    if (!"assay" %in% names(contrast) || is.na(contrast[["assay"]])) contrast[["assay"]] = "RNA"
+    
+    if (contrast[["assay"]] %in% Assays(sc)) {
+      contrast[["use_reduction"]] = FALSE
+    } else if (contrast[["assay"]] %in% Reductions(sc)) {
+      contrast[["use_reduction"]] = TRUE
+    } else {
+      c(error_messages, paste0("The 'assay' column value '",contrast[["assay"]], "' in row ", i, " of the deg contrasts table is neither a Seurat assay nor a Seurat reduction."))
+    }
+    
+    # slot
+    if (!"slot" %in% names(contrast) || is.na(contrast[["slot"]])) contrast[["slot"]] = "data"
+    if (!contrast[["slot"]] %in% c("counts", "data", "scale.data")) c(error_messages, paste0("The 'slot' column value '",contrast[["assay"]], "' in row ", i, " of the deg contrasts table must be 'counts', 'data' or 'scale.data'."))
+    
+    # latent_vars
+    if (!"latent_vars" %in% names(contrast) || is.na(contrast[["latent_vars"]])) {
+      
+      if (length(latent_vars)>0) contrast[["latent_vars"]] = latent_vars else contrast[["latent_vars"]] = NULL
+    } else {
+      contrast[["latent_vars"]] = SplitSpecificationString(contrast[["subset_group"]], first_level_separator=";")
+    }
+    
+    if (!is.null(contrast[["latent_vars"]]) && length(contrast[["latent_vars"]])>0) {
+      if (any(!contrast[["latent_vars"]] %in% colnames(cell_metadata))) {
+        error_messages = c(error_messages, paste("At least one value of column 'latent_vars' in row",i,"of the deg contrasts table or a global latent var cannot be found cell metadata column."))
+      }
+    } 
+    
+    # add error messages
+    contrast = c(contrast, list(error_messages = error_messages))
+    
+    # add contrast rows
+    contrast[["contrast_row"]] = i
+    
+    return(contrast)
+  })
+  
+  
+  # Expand subsets so that there is now one row per subset
```

Comment:
Well for example in param$deg_contrast you could specify:

`condition_group="Cellcycle"; condition_column1="G1"; condition_column="S"; subset_column="seurat_clusters";subset_group="1,2,3,4"` 

This is one row but it means that we have to do four tests (G1 vs S for each cluster). For computation, this one row is converted into four rows:

`condition_group="Cellcycle"; condition_column1="G1"; condition_column="S"; subset_column="seurat_clusters";subset_group="1"` 

`condition_group="Cellcycle"; condition_column1="G1"; condition_column="S"; subset_column="seurat_clusters";subset_group="2"` 

`condition_group="Cellcycle"; condition_column1="G1"; condition_column="S"; subset_column="seurat_clusters";subset_group="3"` 

`condition_group="Cellcycle"; condition_column1="G1"; condition_column="S"; subset_column="seurat_clusters";subset_group="4"` 

Later they are combined again since they logically belong together.

Reply:
Thanks, your explanations are good. Let's see where we add this to the comments.